### PR TITLE
Basic V1

### PR DESCRIPTION
**Pull request by farhan**
### How to test
Simple make the virtual enviroment, install the requirements.txt on the virtual enviroment and run the project.

### Project
This PR implements the project requirements mentioned in the following doc.

https://docs.google.com/document/d/1MXo__IgdqtETBtVIPbxW4i5zWmAv2ab-6Td-prYl8F8/edit?ts=5bdc2608

### Assignment
Create a REST based backend for a recipe management app using Django 2 and Python 3. User stories for the app are as follows:

- New members should be able to register to the app.
- Members should be able to login to the app.
- Members should be able to change their passwords.
- Members should be able to add a new recipe. (Recipe should contain title, brief description, stepwise directions, and ingredients)
- Members should be able to view all the recipes they have created.
- Members should be able to follow other members.
- Members should be able to view recipes of other members they are following.
- Members should be able to edit their profile.
- Members should be able to update or remove their recipes.

### CODE REVIEW

File: core/serializers.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+class UserProfileSerializer(serializers.ModelSerializer):
+    """A serializer for our user profile objects."""
+
+    class Meta:
+        model = models.UserProfile
+        fields = ('id', 'username', 'email', 'first_name', 'password', 'mobile_number')
+        extra_kwargs = {'password': {'write_only': True}}
+
+
+    def create(self, validated_data):
```
Comment (tasawernawaz) [inline core/serializers.py:22]: why we are overriding this method ?

File: core/serializers.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+class FollowerSerializer(serializers.ModelSerializer):
+    """A serializer for our user profile objects."""
+
+    class Meta:
+        model = models.UserProfile
+        fields = ('id', 'username', 'email',)
+
+
+class UserFollowingSerializerList(serializers.ModelSerializer):
```
Comment (tasawernawaz) [inline core/serializers.py]: why we have separate serializer for list ? Tip: if there are two similar serializer you may use inherited serializer.

File: core/urls.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from django.conf.urls import url
+from django.conf.urls import include
+from . import views
```
Comment (tasawernawaz) [inline core/urls.py]: not a good practice of importing with `.`

File: core/views.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+from core import models, serializers, permissions
+from rest_framework import viewsets, status
+from rest_framework.authentication import TokenAuthentication
+
+
+class UserProfileViewSet(viewsets.ModelViewSet):
+    """Handles creating, creating and updating profiles."""
+
+    serializer_class = serializers.UserProfileSerializer
+    queryset = models.UserProfile.objects.all()
```
Comment (tasawernawaz) [inline core/views.py:18]: i guess this is a default queryset, no need to mention it here.

File: core/views.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        return ObtainAuthToken().post(request)
+
+
+class SignUpViewSet(viewsets.ViewSet):
+    """User Sign Up ViewSet."""
+
+    serializer_class = serializers.UserProfileSerializer
+
+    def create(self, request):
+        """Create a new hello message."""
```
Comment (tasawernawaz) [inline core/views.py]: update docstring

File: core/views.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        else:
+            return Response(
+                serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+class UserFollowingViewSet(viewsets.ModelViewSet):
+    """Handles creating, creating and updating profiles."""
+
+    serializer_class = serializers.UserFollowingSerializer
+    queryset = models.UserFollowing.objects.all()
```
Comment (tasawernawaz) [inline core/views.py:85]: same here

File: core/views.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    permission_classes = (permissions.UpdateOwnProfile,)
+
+
+    def perform_create(self, serializer):
+        """Sets the user profile to the logged in user."""
+        serializer.save(following=self.request.user)
+
+
+    def list(self, request):
+        """Return a hello message."""
```
Comment (tasawernawaz) [inline core/views.py]: same here

File: recipe_management/views.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+# Create your views here.
+
+
+def home(request):
+    if request.GET.get('btn_admin'):
+        return redirect('/admin')
+    elif request.GET.get('btn_user_apis'):
+        return redirect('/user-apis')
+    elif request.GET.get('btn_recipe_management_apis'):
+        return redirect('/recipe-management-apis')
```
Comment (tasawernawaz) [inline recipe_management/views.py:15]: did not get this logic ? why we need this.

File: recipe_management_api/serializers.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+            'title',
+            'description',
+            'stepwise_directions',
+            'ingredients',
+            'pictures',
+        )
+        extra_kwargs = {'created_by':{'read_only':True}}
+
+
+class RecipeSerializerList(serializers.ModelSerializer):
```
Comment (tasawernawaz) [inline recipe_management_api/serializers.py]: there isn't much difference between these two serializer, we can use inherited serializer.

File: templates/index.html [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    <meta charset="UTF-8">
+    <title>Recipe Management System</title>
+    <!--<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/meyer-reset/2.0/reset.min.css">-->
+    <!--{% load staticfiles %}-->
+    <link rel="stylesheet" href="style.css">
+    <!--<link rel="stylesheet" href="{% static 'style.css' %}" type="text/css" media="screen" />-->
+    <!--<link href="{% static 'css/style.css' %}?v=1" rel="stylesheet">-->
+    <style>
+        @import url(http://fonts.googleapis.com/css?family=Merriweather);
+*,
```
Comment (tasawernawaz) [inline templates/index.html:14]: move css to its related file

File: templates/index.html [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+<body>
+<h1>Recipe Management System</h1>
+
+<br/><br/><br/>
+
+<form action="" method="get">
+    <input class="my_button" type="submit" name="btn_admin" value="Admin">
+    <input class="my_button" type="submit" name="btn_user_apis" value="User APIs">
+    <input class="my_button" type="submit" name="btn_recipe_management_apis" value="Recipe Management APIs">
```
Comment (tasawernawaz) [inline templates/index.html:100]: instead of this form, you can use simple hyperlinks.

**Review by tasawernawaz — CHANGES_REQUESTED** [review]
Remove unused/commented code
Update relative imports
Update or Remove docstings

**Review by tasawernawaz — CHANGES_REQUESTED** [review]
* Remove/update docstrings
* Change relative imports
* Remove unused/commented code

**Review by tasawernawaz — CHANGES_REQUESTED** [review]
* Remove/update docstrings
* Change relative imports
* Remove unused/commented code

**Review by tasawernawaz — CHANGES_REQUESTED** [review]
* Remove/update docstrings
* Change relative imports
* Remove unused/commented code

**Comment by tasawernawaz** [comment]
@farhan You can integrate [pep8 settings](https://github.com/OrkoHunter/pep8speaks) with you repo to see the violations.

File: core/serializers.py (hunk 1 above)
Comment (farhan) [reply core/serializers.py:22]: I tested without overriding this function.
Without overriding this function we can't create a user successfully.

File: core/serializers.py (hunk 2 above)
Comment (farhan) [reply core/serializers.py]: Optimized the code.

File: core/views.py (hunk 4 above)
Comment (farhan) [reply core/views.py:18]: I tried without this line and got following error.

> AssertionError: `base_name` argument not specified, and could not automatically determine the name from the viewset, as it does not have a `.queryset` attribute.

File: core/views.py (hunk 6 above)
Comment (farhan) [reply core/views.py:85]: same answer

File: recipe_management/views.py (hunk 8 above)
Comment (farhan) [reply recipe_management/views.py:15]: I used this for this redirection scheme for the basic home page redirections.
As I haven't studied the Django framework so far in detail so we can skip these changes for now.

File: templates/index.html (hunk 10 above)
Comment (farhan) [reply templates/index.html:14]: we can skip it in this PR.

File: templates/index.html (hunk 11 above)
Comment (farhan) [reply templates/index.html:100]: we can skip it in this PR.

File: recipe_management_api/serializers.py (hunk 9 above)
Comment (farhan) [reply recipe_management_api/serializers.py]: done!

**Comment by farhan** [comment]
@tasawernawaz its ready for next pass

File: core/admin.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
+from django.contrib import admin
+
+# Register your models here.
```
Comment (tasawernawaz) [inline core/admin.py]: no need of this comment here

File: core/migrations/0001_initial.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
+# Generated by Django 2.1.4 on 2018-12-20 16:14
```
Comment (tasawernawaz) [inline core/migrations/0001_initial.py]: you can remove all migrations and create new initial migration

File: core/models.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
+from django.contrib.auth.models import AbstractUser
+from django.db import models
+
+
+# Create your models here.
```
Comment (tasawernawaz) [inline core/models.py]: same here

File: core/serializers.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+
+# class PasswordChangeSerializer(serializers.ModelSerializer):
+#     """A serializer for our user profile objects."""
+#
+#     confirm_pwd = serializers.CharField()
+#
+#     class Meta:
+#         model = models.UserProfile
+#         fields = ('id', 'password', 'confirm_pwd')
```
Comment (tasawernawaz) [inline core/serializers.py]: remove unused code

File: core/serializers.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+
+    def create(self, validated_data):
+        """Create and return a new user."""
+
+        user = models.UserProfile(
+            username=validated_data['username'],
+            email=validated_data['email'],
+            first_name=validated_data['first_name'],
+            mobile_number=validated_data['mobile_number'],
```
Comment (tasawernawaz) [inline core/serializers.py:29]: is there a chance that `KeyError` can occur.

File: core/urls.py [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+from rest_framework import routers
+from django.urls import path
+
+router = routers.DefaultRouter()
+
+router.register(r'sign-up', views.SignUpViewSet, base_name='sign-up')
+router.register(r'log-in', views.LoginViewSet, base_name='log-in')
+# router.register(r'password-change', views.PasswordChangeViewSet, base_name='password-change')
```
Comment (tasawernawaz) [inline core/urls.py]: remove

File: core/urls.py [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+router.register(r'profile-viewset', views.UserProfileViewSet)
+router.register(r'following-viewset', views.UserFollowingViewSet)
+
+urlpatterns = [
+]
```
Comment (tasawernawaz) [inline core/urls.py:15]: whats the use of it ?

File: core/validators.py [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+from django.core import validators
+
+
+@deconstructible
+class PhoneNumberValidator(validators.RegexValidator):
+    regex = r'^\+?1?\d{9,15}$'
+    message = (
+        "Phone number must be entered in the format: '+999999999'. Up to 15 digits allowed."
+    )
+    flags = 0
```
Comment (tasawernawaz) [inline core/validators.py]: is this used anywhere ?

File: core/views.py [hunk 20]
```diff
@@ -0,0 +1,76 @@
+from django.shortcuts import render
+from rest_framework.authtoken.serializers import AuthTokenSerializer
+from rest_framework.authtoken.views import ObtainAuthToken
+from rest_framework.response import Response
+
+from core import models, serializers, permissions
+from rest_framework import viewsets, status
+from rest_framework.authentication import TokenAuthentication
```
Comment (tasawernawaz) [inline core/views.py:8]: optimize imports

File: core/views.py [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+class UserProfileViewSet(viewsets.ModelViewSet):
+    """Handles creating, creating and updating profiles."""
+
+    serializer_class = serializers.UserProfileSerializer
+    queryset = models.UserProfile.objects.all()
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (permissions.UpdateOwnProfile,)
+    # filter_backends = (filters.SearchFilter,)
+    # search_fields = ('name', 'email',)
```
Comment (tasawernawaz) [inline core/views.py]: remove

File: recipe_management_api/models.py [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class Recipe(models.Model):
+    """Basic Recipe model class"""
+    created_by = models.ForeignKey('core.UserProfile', on_delete=models.CASCADE)
+    title = models.CharField(max_length=100)
+    description = models.CharField(max_length=10000)
+    stepwise_directions = models.CharField(max_length=1000)
+    ingredients = models.CharField(max_length=1000)
+    pictures = models.ImageField(default=None)
```
Comment (tasawernawaz) [inline recipe_management_api/models.py]: nit: picture

**Review by tasawernawaz — CHANGES_REQUESTED** [review]

File: core/serializers.py (hunk 15 above)
Comment (farhan) [reply core/serializers.py]: I had already removed them, you might don't update the code

File: core/validators.py (hunk 19 above)
Comment (farhan) [reply core/validators.py]: nops, removed

File: core/serializers.py (hunk 16 above)
Comment (farhan) [reply core/serializers.py:29]: I have read from here about KeyError, I didn't get where is the change of occurrence of KeyError.
https://docs.python.org/3/library/exceptions.html#KeyError

**Comment by farhan** [comment]
@tasawernawaz  I have addressed the change requests, please have another round of review.
And as its an assignment project so you can skip the docs part in it ;)

File: core/urls.py (hunk 18 above)
Comment (tasawernawaz) [reply core/urls.py:15]: ?

File: core/serializers.py (hunk 16 above)
Comment (tasawernawaz) [reply core/serializers.py:29]: e.g if we don't have `mobile_number` in `validated_date`, it will throw `KeyError`

**Review by tasawernawaz — APPROVED** [review]
just couple of comments otherwise looks good now.

File: core/urls.py (hunk 17 above)
Comment (farhan) [reply core/urls.py]: removed

File: core/urls.py (hunk 18 above)
Comment (farhan) [reply core/urls.py:15]: its removed

File: core/serializers.py (hunk 16 above)
Comment (farhan) [reply core/serializers.py:29]: This is how validation is working on the api at this moment
Screen Shot 2019-08-08 at 8 14 40 PM
Screen Shot 2019-08-08 at 8 14 58 PM

**Comment by farhan** [comment]
> just couple of comments otherwise looks good now.

@tasawernawaz  comments addressed, next pass please

**Pull request merged**